File: src/main/java/net/asodev/islandutils/util/Scheduler.java
package net.asodev.islandutils.util;

import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.minecraft.client.Minecraft;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;

public class Scheduler {
    private final List<Task> tasks = new ArrayList<>();

    public Scheduler() {
        ClientTickEvents.END_CLIENT_TICK.register((client) -> {
            if (tasks.isEmpty()) return;
            tasks.removeIf(s -> s.tick(client));
        });
    }

    static class Task {
        private boolean shouldRemove = false;
        private int ticks;
        private final Consumer<Minecraft> callback;

        public Task(int ticks, Consumer<Minecraft> callback) {
            this.ticks = ticks;
            this.callback = callback;
        }

        public boolean tick(Minecraft client) {
            if (shouldRemove) return true;

            ticks--;
            if (ticks > 0) return false;
            callback.accept(client);
            return true;
        }

        public void cancel() {
            shouldRemove = true;
        }
    }

    private static Scheduler instance;

    public static Task schedule(int afterTicks, Consumer<Minecraft> callback) {
        Task task = new Task(afterTicks, callback);
        instance.tasks.add(task);
        return task;
    }
    public static Task nextTick(Consumer<Minecraft> callback) {
        return schedule(1, callback);
    }
    public static void create() {
        instance = new Scheduler();
    }

}

File: src/main/kotlin/dev/asodesu/islandutils/api/Scheduler.kt
package dev.asodesu.islandutils.api

import dev.asodesu.islandutils.api.Scheduler.Task
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.minecraft.client.Minecraft
import kotlin.time.Duration

/**
 * A simple tick scheduler
 */
object Scheduler {
    private val tasks = mutableListOf<Task>()
    fun init() {
        ClientTickEvents.END_CLIENT_TICK.register {
            this.tick(it)
        }
    }

    private fun tick(client: Minecraft) {
        if (tasks.isEmpty()) return

        tasks.removeIf { it.tick(client) }
    }

    /**
     * Schedule a new task to execute in the specified duration
     *
     * @param duration The amount of time to wait
     * @param callback The task to execute
     */
    fun runAfter(duration: Duration, callback: (Minecraft) -> Unit): Task {
        return Task(duration, callback)
            .also { tasks += it }
    }

    class Task(duration: Duration, private val callback: (Minecraft) -> Unit) {
        private var remove = false
        private var ticks = duration.ticks

        fun tick(client: Minecraft): Boolean {
            if (remove) return true

            if (--ticks > 0) return false
            callback(client)
            return true
        }

        fun cancel() {
            remove = true
        }
    }
}